cmake/modules/RootMacros.cmake

```diff
@@ -283,115 +283,226 @@ function(ROOT_GENERATE_DICTIONARY dictionary)
     set(libprefix "")
   endif()
 
-  #---Get the list of include directories------------------
-  get_directory_property(incdirs INCLUDE_DIRECTORIES)
-  # rootcling invoked on foo.h should find foo.h in the current source dir,
-  # no matter what.
-  list(APPEND incdirs ${CMAKE_CURRENT_SOURCE_DIR})
+  if((CMAKE_PROJECT_NAME STREQUAL ROOT) AND (TARGET ${ARG_MODULE}))
+    set(incdirs)
+
+    if (cxxmodules OR runtime_cxxmodules)
+       # Comments from Vassil:
+       # FIXME: We prepend ROOTSYS/include because if we have built a module
+       # and try to resolve the 'same' header from a different location we will
+       # get a redefinition error.
+       # We should remove these lines when the fallback include is removed. Then
+       # we will need a module.modulemap file per `inc` directory.
+       list(APPEND incdirs ${CMAKE_BINARY_DIR}/include)
+    else()
+       # Comments from Sergey: because of TVirtualObject.h, appearing in each dictionary
+       list(APPEND incdirs ${CMAKE_SOURCE_DIR}/io/io/inc)
+    endif()
+
+    set(headerdirs)
 
-  if(TARGET ${ARG_MODULE})
     get_target_property(target_incdirs ${ARG_MODULE} INCLUDE_DIRECTORIES)
-    foreach(dir ${target_incdirs})
-      string(REGEX REPLACE "^[$]<BUILD_INTERFACE:(.+)>" "\\1" dir ${dir})
-      if(NOT ${dir} MATCHES "^[$]")
-        list(APPEND incdirs ${dir})
-      endif()
+    if(target_incdirs)
+       foreach(dir ${target_incdirs})
+          string(REGEX REPLACE "^[$]<BUILD_INTERFACE:(.+)>" "\\1" dir ${dir})
+          if(NOT ${dir} MATCHES "^[$]")
+             list(APPEND incdirs ${dir})
+          endif()
+          if(${dir} MATCHES "^${CMAKE_SOURCE_DIR}")
+             list(APPEND headerdirs ${dir})
+          endif()
+       endforeach()
+    endif()
+
+
+    # Comments from Sergey
+    # FIXME: we have to exclude all source directories from includes while they duplicated in ${CMAKE_BINARY_DIR}/include
+    # rootcling is not able to handle such situation, therefore ${CMAKE_BINARY_DIR}/include is used as source for ROOT headers
```

Comment:
What's wrong with rootcling? Do you mean it cannot handle paths to the concrete inc dirs along with -I$ROOTSYS/include?

Reply:
I just repeat your own comments. 
Due to module map rootcling already has full list of includes and it thinks - all includes should come from `$ROOTSYS/include`. Therefore if include with similar names found via source dir - `cling` immediately complains. Only solution for now - fully remove source dirs from include paths when producing dictionaries